### ISSUE TITLE
PvE battles :suggestion:

### ISSUE THREAD

**Issue by Kerodas** [naruto-unison/naruto-unison#41]
I suggest adding PvE battles option. The same as practice mode, but against random teams. PvE battles will not have daily rewards and will reward half the amount of currency that PvP battles give, but they could be done for unlocking characters on the current stage of game development, which will also allow us to test missions and their design.

In the future PvE battles might have difficulty selection as well:
Easy - Enemy characters have 100 HP
Advanced - Enemy characters have 200 HP
Hard - Enemy characters have 300 HP
This adds artificial difficulty without having to make 3 different AI behavior patterns. 

Random daily mission could also be added for this mode in the future, which will be a mission that will reward you for winning some matches with certain character or in certain conditions, like win the battle with only 1 character remaining. (similar to a hearthstone daily tasks).

You could also add PvE records and ranks for the game which are separate from PvP ones and which don't count for ladderboards. This should attract people who are into single player games as well.

Characters unlocked in PvE mode can be played only in PvE mode, but they will be flagged for being available in PvP mode with currency which is obtained only in PvP mode. So if a person has unlocked a character in PvE mode, he can spend like 500 coins that are obtained by winning 10 games in PvP. This way, people who are interested in PvE only content should not be pressured to do PvP, while people who do both PvE and PvP should not be pressured to do only PvP battles because of the rewards.

**Comment by jnbooth** [comment]
Yep, this is in the works for as soon as the mission system is finished, which should be by the end of the week, and then the Wars system, which honestly should only take a day or two. I love the idea of PvE-specific ranks and unlocks. The DNA system is already implemented so I'm wary of adding another currency, but I will definitely play around with that.

> they could be done for unlocking characters on the current stage of game development, which will also allow us to test missions and their design

Ah, I should add a note to the README. Mission locking is enabled by editing [config/settings.yml](https://github.com/naruto-unison/naruto-unison/blob/master/config/settings.yml) and uncommenting this line:
```yaml
# unlock-all: false
```
to
```yaml
unlock-all: false
```
By default, `unlock-all` is true in `yesod devel` and false in `yesod keter`.

A few minor differences, subject of course to changes:
- Rather than giving enemies more HP, higher difficulty enemies will have special percentage damage reduction that cannot be pierced. It's more consistent that way. Giving the hardest of the hard fights 80% damage reduction is appealing to me because skill damage is generally multiples of 5 anyway, but that might be a little too difficult/too much of a slog.
- Rather than completely random matchups, PvE teams will work as random selections within pools of characters. So for example, some pools might be only 3 characters (i.e. Gaara, Temari, Kankuro), but others might be much larger (e.g. 3 random Swordsmen of the Mist). Kind of like how Battle Tree missions work in Pokémon Gen VII. It's a small difference that adds some flavor. Obviously there will be a lot of pools, and admins will be able to add more without having to modify the code by using the Admin page (localhost:3000/admin in `yesod devel`).
- Randomized boss fights will add additional buffs like giving the boss unlimited chakra, destructible defense every turn, etc. Like the random boss buffs in Diablo.
- There will also be scripted boss fights. I think it would be really cool to have a Sage Mode Kabuto fight where he resurrects random Reanimated characters and is invulnerable until you kill some amount of them, a Shukaku Gaara fight where he starts transformed and you can only use Team 7, a fight against an army of White Zetsu, etc. The problem here is I don't know how to do this without requiring regular modification of the codebase. Maybe that's okay, but it irks me.

**Comment by Kerodas** [comment]
Damage reduction might not work due to affliction damage.

Battle trees should work even better than random team generator.

Randomized boss fights sounds like a cool idea, might as well make boss rush mode after.

I think you should concentrate on boss fights once everything else is finished. Making PvE mode with battle trees should help with that, since it's basically a separate instance which targets a specific group of character during battle creation. Following the same pattern, you could make an instance with a boss character, which follows AI commands every turn, like you scripted team 7 for practice mode. Make boss character's skills cost nothing and force AI to make a certain move every turn. That's a basic concept at least.

**Comment by jnbooth** [comment]
That was what I meant by "cannot be pierced." Damage reduction that also affects piercing and affliction damage. It's what Utakata (S) has from _Six-Tailed Transformation_.

The objective column is the index within the objective list of the mission. So if it's 0, it's the first objective. If it's 1, it's the second objective. I think all the missions currently have four objectives, so there should be 0 through 3 for each.
